### PR TITLE
can scroll to last few lines of text editor when error list is open

### PR DESCRIPTION
**Pull request by hristoramirez**
fixed the issue where the first or last few lines of code would be unreachable when scrolling

fixes microsoft/pxt-microbit#2824

### CODE REVIEW

File: webapp/src/errorList.tsx [hunk 1]
```diff
@@ -43,30 +43,26 @@ export class ErrorList extends React.Component<ErrorListProps, ErrorListState> {
...
-            {!isCollapsed &&
-                <div className="errorListInner">
-                    {errorListContent}
+                }
+            </div> )
+
+        return (
+            <div className={`errorList ${isCollapsed ? 'errorListSummary' : ''}`} hidden={!errorsAvailable}>
+                <div className="errorListHeader" role="button" onClick={this.onCollapseClick}>
+                    <h4>Problems</h4>
```
Comment (pelikhan) [inline webapp/src/errorList.tsx:59]: Problems -> {lf(“Problems”)}

File: webapp/src/errorList.tsx [hunk 2]
```diff
@@ -43,30 +43,26 @@ export class ErrorList extends React.Component<ErrorListProps, ErrorListState> {
...
-        </div>
-
-        const errorListContent = <div className="ui selection list">
-            {
-                (errors).map((e, index) =>
+        const errorListContent = (
+            <div className="ui selection list">
+                {(errors).map((e, index) =>
                     <div className="item" key={errorKey(e)} role="button" onClick={createOnErrorMessageClick(e, index)}>
                         {lf("Line {0}: ", e.line + 1)}
```
Comment (pelikhan) [inline webapp/src/errorList.tsx:50]: {lf("Line {0}: {1}", e.line + 1, message)}

File: webapp/src/errorList.tsx [hunk 3]
```diff
@@ -43,30 +43,26 @@ export class ErrorList extends React.Component<ErrorListProps, ErrorListState> {
...
-                <div className="errorListInner">
-                    {errorListContent}
+                }
+            </div> )
+
+        return (
+            <div className={`errorList ${isCollapsed ? 'errorListSummary' : ''}`} hidden={!errorsAvailable}>
+                <div className="errorListHeader" role="button" onClick={this.onCollapseClick}>
+                    <h4>Problems</h4>
+                    <div className="ui red circular label countBubble">{errors?.length}</div>
```
Comment (pelikhan) [inline webapp/src/errorList.tsx:60]: errors is guaranteed to be defined here or it would have crashed earlier.

**Comment by pelikhan** [comment]
I'm not sure i understand how this issue got fixed. Could you explain a bit more in the comments?

File: webapp/src/errorList.tsx [hunk 4]
```diff
@@ -43,30 +43,26 @@ export class ErrorList extends React.Component<ErrorListProps, ErrorListState> {
...
-        const errorListHeader = <div className="errorListHeader" role="button" onClick={this.onCollapseClick}>
-            <h4>Problems</h4>
-            <div className="ui red circular label countBubble">{errors?.length}</div>
-            <div className="toggleButton"><sui.Icon icon={`chevron ${isCollapsed ? 'up' : 'down'}`} onClick={this.onCollapseClick} /></div>
-        </div>
-
-        const errorListContent = <div className="ui selection list">
-            {
-                (errors).map((e, index) =>
+        const errorListContent = (
```
Comment (pelikhan) [inline webapp/src/errorList.tsx:46]: don't compute UI elements eagerly, compute them if not collapsed

File: webapp/src/monaco.tsx [hunk 5]
```diff
@@ -728,6 +728,9 @@ export class Editor extends toolboxeditor.ToolboxEditor {
             });
 
             if (monacoToolboxDiv) monacoToolboxDiv.style.height = `100%`;
+
+            // ensures all lines are displayed
+            this.editor.focus();
```
Comment (pelikhan) [inline webapp/src/monaco.tsx:733]: Does this mean that the monaco editor will steal focus when the user resizes the window?

Comment (darzu) [reply webapp/src/monaco.tsx:733]: Great question, we don't want to do that. Hristo, we might need to dig into .focus() to see what it's actually doing that fixes the sizing issue and maybe only call a subset of focus()'s behavior.

Comment (darzu) [reply webapp/src/monaco.tsx:733]: Also it's worth testing this with tab navigation. If you're tabbing around the page and this triggers does  the user lose their tab location?

Comment (hristoramirez) [reply webapp/src/monaco.tsx:733]: Ok I'll look into this more to see if there is another way

Comment (jwunderl) [reply webapp/src/monaco.tsx:733]:
```suggestion
            this.editor.layout();
```

@hristoramirez could you check if this fixes the problem? at a glance it looks like the issue is that the dimensions and offset are hardcoded on each of the divs (e.g. `div.monaco-scrollable-element` that's a parent of the individual lines), and that gets re-evaluated on focus: https://github.com/microsoft/monaco-editor/blob/master/monaco.d.ts#L2133

Comment (hristoramirez) [reply webapp/src/monaco.tsx:733]: @jwunderl I tried it, but it gets stuck in a loop while loading

Comment (hristoramirez) [reply webapp/src/monaco.tsx:733]: @jwunderl so far it looks like it fixes the problem

Comment (darzu) [reply webapp/src/monaco.tsx]: Might be worth checking that this doesn't slow down the editor resize speed

**Pull request merged**